_source/_posts/2021-09-12-spring-boot-spa.md

```diff
@@ -0,0 +1,686 @@
+
+```bash
+./mvnw spring-boot:run
+```
+
+You should get a lot of console output that ends with the following.
+
+```bash
+2021-09-01 11:49:42.136  INFO 370077 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat started on port(s): 8082 (http) with context path ''
+2021-09-01 11:49:42.143  INFO 370077 --- [           main] com.okta.springbootspa.DemoApplication   : Started DemoApplication in 2.021 seconds (JVM running for 2.199)
+```
+
+## Use The Vue CLI To Bootstrap The Client App
+
+The next step is to build the front-end, client application. The application will use Vue and the Okta Sign-In Widget to implement a secure client application. The Okta Sign-In Widget allows you to easily add Single Sign-On (SSO) and a secure login form to any application.
+
+The instructions below for integrating the Okta Sign-In widget with Vue are largely pulled from [this article](https://developer.okta.com/code/vue/okta_vue_sign-in_widget/#add-an-openid-connect-client-in-okta) on the Okta website, which is a great reference if you have any questions.
+
+Create the app using the Vue CLI. Run the following command from the project root directory (the directory above the spring boot project directory). It will create a new directory, `okta-app`, that will contain the client application.
+
+```bash
+vue create okta-app
+```
+
+Select `Manually select features`.
+
+Make sure `Router` and `Choose Vue version` are selected. `Babel` and `Linter / Formatter` were automatically selected for me as well, which is fine.
+
+Select version `3.x`. 
+
+Enable `history mode for router` (type `y`).
+
+Select `ESLint with error prevention only`.
+
+The defaults for the remainder of the options are fine: lint on save, saving configuration in dedicated files, and not saving this as a preset.
+
+Navigate into the project directory.
+
+```bash
+cd okta-app
+```
+
+You need to install a few dependencies. The first three are required for Okta security and the sign-in widget. The last, `axios`, is the library you'll use to make the HTTP request to the resource server.
+
+```bash
+npm install --save \
+@okta/okta-signin-widget \
+@okta/okta-vue \
+@okta/okta-auth-js \
+axios
+```
+
+Create a `src/okta/index.js` file:
+
+```vue
+import OktaSignIn from '@okta/okta-signin-widget'
+import { OktaAuth } from '@okta/okta-auth-js'
+
+const yourOktaUri = '${yourOktaUri}';
+const clientId = '${clientId}';
+
+const oktaSignIn = new OktaSignIn({
+    baseUrl: `${yourOktaUri}`,
+    clientId: clientId,
+    redirectUri: 'http://localhost:8080/login/callback',
+    authParams: {
+        pkce: true,
+        issuer: `${yourOktaUri}/oauth2/default`,
+        display: 'page',
+        scopes: ['openid', 'profile', 'email']
+    }
+});
+
+const oktaAuth = new OktaAuth({
+    issuer: `${yourOktaUri}/oauth2/default`,
+    clientId: clientId,
+    redirectUri: window.location.origin + '/login/callback',
+    scopes: ['openid', 'profile', 'email']
+})
```

Comment:
See the above comment about the redirect/"okta-hosted-login" example, but that might help DRY this config block out a little too:
https://github.com/okta/samples-js-vue/blob/master/okta-hosted-login/src/config.js#L5-L9